mantidproject/mantid
# New algorithm to refine lattice parameters across runs with separate U matrix/orientation

**Description of work.**

New algorithm `FindGlobalBMatrix` to refine the lattice parameters (encoded in the B matrix) across many runs with separate U matrix/orientation which is useful for when the goniometer is inaccurate/not well known.

The quantity minimised is the average of the square of the difference in `QSample` between peak the peak at integer HKL (the minimisation is performed by scipy `leastsq` function). The average rather than the sum of the squared residuals is used so as not to penalise the indexing of more peaks as the UB matrix becomes more accurate.

`FindGlobalBMatrix` adds a different UB to each peak workspace that can be indexed (the algorithm ensures consistent indexing across runs - e.g. not swapped `H<->K` and `L->-L`) - however all the workspaces have the same lattice parameters (it's just U that's different between runs)

**To test:**

Use some SXD data of cubic NaCl (a=b=c=6.65, alpha=beta=gamma = 90) and see how close the refined parameters are to nominal

(1) Run this script to load the data and find peaks
```
runs = list(range(32864, 32866, 1))

def process_run(runno):
    w = Load(Filename='SXD' + str(runno) +'.raw', OutputWorkspace='SXD' + str(runno))
    LoadParameterFile(Workspace='SXD' + str(runno), 
        Filename=r'\\olympic\Babylon5\Public\RWaite\SXD32863_calibration.xml') # updated calibration
    NormaliseByCurrent(InputWorkspace='SXD' + str(runno), OutputWorkspace='SXD' + str(runno))
    CropWorkspace(InputWorkspace='SXD' + str(runno), OutputWorkspace='SXD' + str(runno), XMin=1000, XMax=10000)
    # SetGonio doesn't recognise wccr log so have to manually average 
    wccr = w.run().getPropertyAsSingleValueWithTimeAveragedMean('wccr')
    SetGoniometer(Workspace='SXD' + str(runno), Axis0=str(wccr)+',0,1,0,1') 
    # Find peaks
    FindSXPeaks(InputWorkspace='SXD' + str(runno), PeakFindingStrategy='AllPeaks', AbsoluteBackground=5,
        ResolutionStrategy='AbsoluteResolution', XResolution=400, PhiResolution=3, TwoThetaResolution=3, 
        OutputWorkspace='SXD' + str(runno) +'_peaks')
    FilterPeaks(InputWorkspace='SXD' + str(runno) +'_peaks', OutputWorkspace='SXD' + str(runno) +'_peaks', 
        Criterion='!=', BankName='bank1')  # remove peaks in Bank 1 as calibration poor

for irun, runno in enumerate(runs):
    process_run(runno)
    if irun == 0:
        AllPeaks = CloneWorkspace(InputWorkspace='SXD' + str(runno) + '_peaks')
    else:
        AllPeaks = CombinePeaksWorkspaces(AllPeaks, 'SXD' + str(runno) + '_peaks',  CombineMatchingPeaks=False)
```
 (2) Run this to refine the global B matrix (it is slowed down by the call to `FindUBUsingLatticeParameters` to find the initial UB)
```
FindGlobalBMatrix(PeakWorkspaces = ['SXD32864_peaks', 'SXD32865_peaks'],
    a=5.4, b=5.8,c=5.9,alpha=88,beta=88,gamma=89)
```
It should print this in the log  - format is  [ a, b, c, alpha, beta, gamma]
```
Lattice Parameters: [ 5.653631  5.65586   5.656385 89.99074  90.077986 90.003773]
Parameter Errors  : [0.007926 0.011577 0.013786 0.181732 0.176361 0.132551]
```
(3) Play around with the initial parameters - it should return the same result (typically we know the rough lattice parameter within 0.1 Ang so already the initial guess used above is quite an extreme example)

(4) Compare this to running `FindUBUsingLatticeParameters` with a much better guess on the combined workspaces which has a single orientation for all runs that is rotated by the goniometer matrix.
```
FindUBUsingLatticeParameters(PeaksWorkspace='AllPeaks', a=5.65, b=5.65, c=5.65, alpha=90, beta=90, gamma=90)
```
It should print 
```
Lattice Parameters:    5.641454    5.697618    5.670424   89.542192   89.929136   90.329593   182.254616
Parameter Errors  :    0.014173    0.017662    0.008407    0.196720    0.167132    0.228287    0.775815
```
You will see the lattice parameters are less consistent with cubic NaCl lattice. 



Fixes #31909 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
